File: beanmachine/graph/nmc_test.cpp
// Copyright (c) Facebook, Inc. and its affiliates.
#include <gtest/gtest.h>
#include <random>

#include "beanmachine/graph/graph.h"

using namespace beanmachine::graph;

TEST(testnmc, beta_binomial) {
  Graph g;
  uint a = g.add_constant_pos_real(5.0);
  uint b = g.add_constant_pos_real(3.0);
  uint n = g.add_constant((natural_t)10);
  uint beta = g.add_distribution(
      DistributionType::BETA,
      AtomicType::PROBABILITY,
      std::vector<uint>({a, b}));
  uint prob = g.add_operator(OperatorType::SAMPLE, std::vector<uint>({beta}));
  uint binomial = g.add_distribution(
      DistributionType::BINOMIAL,
      AtomicType::NATURAL,
      std::vector<uint>({n, prob}));
  uint k = g.add_operator(OperatorType::SAMPLE, std::vector<uint>({binomial}));
  g.observe(k, (natural_t)8);
  g.query(prob);
  // Note: the posterior is p ~ Beta(13, 5).
  int num_samples = 10000;
  std::vector<std::vector<AtomicValue>> samples =
      g.infer(num_samples, InferenceType::NMC);
  double sum = 0;
  double sumsq = 0;
  for (const auto& sample : samples) {
    const auto& s = sample.front();
    ASSERT_EQ(s.type, AtomicType::PROBABILITY);
    sum += s._double;
    sumsq += s._double * s._double;
  }
  double mean = sum / num_samples;
  double var = sumsq / num_samples - mean * mean;
  EXPECT_NEAR(mean, 13.0 / 18.0, 1e-2);
  EXPECT_NEAR(var, 13.0 * 5.0 / (18.0 * 18.0 * 19.0), 1e-3);
}

TEST(testnmc, net_norad) {
  // The NetNORAD model is as follows:
  // We have a number of components s.t. the i th one has a drop rate drop_i
  // We observe packets sent on multiple paths criss-crossing these components
  // For each path we observe the number sent and number dropped/received
  // We model the drop rate of a path as `1 - product_i (1 - drop_i)` for i in
  // path And we are trying to infer drop_i for each component The prior on
  // `drop_i ~ Beta(.0001, 100)` i.e. 1 in million odds The prior on `pkts
  // dropped on a path ~ Binomial(pkts sent, path-drop-rate)`
  Graph g;
  uint a = g.add_constant_pos_real(0.0001);
  uint b = g.add_constant_pos_real(100.0);
  uint drop_prior = g.add_distribution(
      DistributionType::BETA, AtomicType::PROBABILITY, std::vector<uint>{a, b});
  std::vector<uint> comp_rates; // complement of the drop rates
  for (int i = 0; i < 4; i++) {
    uint drop =
        g.add_operator(OperatorType::SAMPLE, std::vector<uint>{drop_prior});
    g.query(drop);
    uint comp =
        g.add_operator(OperatorType::COMPLEMENT, std::vector<uint>{drop});
    comp_rates.push_back(comp);
  }
  // for each path the pkts sent, pkts recvd, component ids
  std::vector<std::tuple<uint, uint, std::vector<uint>>> paths = {
      {200, 200, {0, 1}},
      {200, 180, {1, 2}},
      {200, 170, {2, 3}},
      {200, 199, {0, 1, 3}}};
  for (const auto& path : paths) {
    uint pkts_sent = g.add_constant((natural_t)std::get<0>(path));
    std::vector<uint> path_comp_rates;
    for (uint id : std::get<2>(path)) {
      path_comp_rates.push_back(comp_rates[id]);
    }
    uint prod = g.add_operator(OperatorType::MULTIPLY, path_comp_rates);
    // path_drop_rate = 1 - product_{i | i in path} (1 - drop_i)
    uint path_drop_rate =
        g.add_operator(OperatorType::COMPLEMENT, std::vector<uint>{prod});
    uint path_dist = g.add_distribution(
        DistributionType::BINOMIAL,
        AtomicType::NATURAL,
        std::vector<uint>{pkts_sent, path_drop_rate});
    uint pkts_dropped =
        g.add_operator(OperatorType::SAMPLE, std::vector<uint>{path_dist});
    g.observe(pkts_dropped, (natural_t)(std::get<0>(path) - std::get<1>(path)));
  }
  const std::vector<double>& means = g.infer_mean(10000, InferenceType::NMC);
  // component 0 is less than 1 because it has fewer dropped packets
  EXPECT_LT(means[0], means[1]);
  // 1 is less than 3 because it has fewer dropped packets
  EXPECT_LT(means[1], means[3]);
  // 2 is 100x of 1 because all paths involving 2 have high dropped packets
  EXPECT_LT(10 * means[1], means[2]);
}

TEST(testnmc, normal_normal) {
  Graph g;
  auto mean0 = g.add_constant(0.0);
  auto sigma0 = g.add_constant_pos_real(5.0);
  auto dist0 = g.add_distribution(
      DistributionType::NORMAL,
      AtomicType::REAL,
      std::vector<uint>{mean0, sigma0});
  auto x = g.add_operator(OperatorType::SAMPLE, std::vector<uint>{dist0});
  auto x_sq = g.add_operator(OperatorType::MULTIPLY, std::vector<uint>{x, x});
  auto sigma1 = g.add_constant_pos_real(10.0);
  auto dist1 = g.add_distribution(
      DistributionType::NORMAL, AtomicType::REAL, std::vector<uint>{x, sigma1});
  auto y = g.add_operator(OperatorType::SAMPLE, std::vector<uint>{dist1});
  g.observe(y, 100.0);
  g.query(x);
  g.query(x_sq);
  const std::vector<double>& means = g.infer_mean(10000, InferenceType::NMC);
  // posterior of x is N(20, sqrt(20))
  EXPECT_NEAR(means[0], 20, 0.1);
  EXPECT_NEAR(means[1] - means[0] * means[0], 20, 1.0);
}

TEST(testnmc, flat_normal) {
  // This test learns both the mean and standard deviation
  Graph g;
  auto mean_prior = g.add_distribution(
      DistributionType::FLAT, AtomicType::REAL, std::vector<uint>{});
  auto mean =
      g.add_operator(OperatorType::SAMPLE, std::vector<uint>{mean_prior});
  auto sigma_prior = g.add_distribution(
      DistributionType::FLAT, AtomicType::POS_REAL, std::vector<uint>{});
  auto sigma =
      g.add_operator(OperatorType::SAMPLE, std::vector<uint>{sigma_prior});
  auto likelihood = g.add_distribution(
      DistributionType::NORMAL,
      AtomicType::REAL,
      std::vector<uint>{mean, sigma});
  // we will sample some values from the normal distribution and try to infer
  // the empirical mean and std
  int num_obs = 20;
  std::normal_distribution<double> dist(4.3, 5.1);
  std::mt19937 gen(314521);
  double emp_mean = 0;
  double emp_var = 0;
  for (int i = 0; i < num_obs; i++) {
    double val = dist(gen);
    auto n =
        g.add_operator(OperatorType::SAMPLE, std::vector<uint>{likelihood});
    g.observe(n, val);
    emp_mean += val / num_obs;
    emp_var += val * val / (num_obs - 1);
  }
  emp_var -= emp_mean * emp_mean * num_obs /
      (num_obs - 1); // unbiased estimator of sample variance
  g.query(mean);
  g.query(sigma);
  const std::vector<double>& post_means =
      g.infer_mean(10000, InferenceType::NMC);
  EXPECT_NEAR(post_means[0], emp_mean, 0.1);
  EXPECT_NEAR(post_means[1], std::sqrt(emp_var), 1.0);
}

TEST(testnmc, gmm_two_components) {
  // Test two component Gaussian Mixture Model
  // tau ~ Beta(1, 1)
  // mu_{false|true} ~ Normal(0, 100)
  // c_i ~ Bernoulli(tau)
  // X_i ~ Normal(mu_{c_i}, 1) are observed with values [1.0, 8.0, 11.0]
  // and observe c_0 = false (to break symmetry)
  // posterior of mu_false will be close to 1 and
  // posterior of mu_true will be close to 9.5
  // posterior of tau is Beta (3, 2) which has a mean of 0.6
  Graph g;
  auto zero = g.add_constant(0.0);
  auto one = g.add_constant_pos_real(1.0);
  auto hundred = g.add_constant_pos_real(100.0);
  auto tau_prior = g.add_distribution(
      DistributionType::BETA,
      AtomicType::PROBABILITY,
      std::vector<uint>{one, one});
  auto tau = g.add_operator(OperatorType::SAMPLE, std::vector<uint>{tau_prior});
  auto c_prior = g.add_distribution(
      DistributionType::BERNOULLI, AtomicType::BOOLEAN, std::vector<uint>{tau});
  auto mu_prior = g.add_distribution(
      DistributionType::NORMAL,
      AtomicType::REAL,
      std::vector<uint>{zero, hundred});
  auto mu_false =
      g.add_operator(OperatorType::SAMPLE, std::vector<uint>{mu_prior});
  auto mu_true =
      g.add_operator(OperatorType::SAMPLE, std::vector<uint>{mu_prior});
  const std::vector<double> X_DATA = {1.0, 8.0, 11.0};
  for (const auto x_val : X_DATA) {
    auto c_i = g.add_operator(OperatorType::SAMPLE, std::vector<uint>{c_prior});
    if (x_val == 1.0) {
      g.observe(c_i, false);
    }
    auto mu = g.add_operator(
        OperatorType::IF_THEN_ELSE, std::vector<uint>{c_i, mu_true, mu_false});
    auto x_prior = g.add_distribution(
        DistributionType::NORMAL, AtomicType::REAL, std::vector<uint>{mu, one});
    auto x = g.add_operator(OperatorType::SAMPLE, std::vector<uint>{x_prior});
    g.observe(x, x_val);
  }
  g.query(tau);
  g.query(mu_false);
  g.query(mu_true);
  const std::vector<double>& post_means =
      g.infer_mean(1000, InferenceType::NMC);
  EXPECT_NEAR(post_means[0], 0.6, 0.01);
  EXPECT_NEAR(post_means[1], 1.0, 0.1);
  EXPECT_NEAR(post_means[2], 9.5, 0.1);
}

File: beanmachine/graph/nmc_test.py
# Copyright (c) Facebook, Inc. and its affiliates.
import math
import unittest

import numpy as np
from beanmachine import graph


class TestNMC(unittest.TestCase):
    # see https://www.jstatsoft.org/article/view/v012i03/v12i03.pdf
    def test_eight_schools(self):
        # For each school, the average treatment effect and the standard deviation
        DATA = [
            (28.39, 14.9),
            (7.94, 10.2),
            (-2.75, 16.3),
            (6.82, 11.0),
            (-0.64, 9.4),
            (0.63, 11.4),
            (18.01, 10.4),
            (12.16, 17.6),
        ]
        # the expected mean and standard deviation of each random variable
        EXPECTED = [
            (11.1, 9.1),
            (7.6, 6.6),
            (5.7, 8.4),
            (7.1, 7.0),
            (5.1, 6.8),
            (5.7, 7.3),
            (10.4, 7.3),
            (8.3, 8.4),
            (7.6, 5.9),  # overall mean
            (6.7, 5.6),  # overall std
        ]
        g = graph.Graph()
        zero = g.add_constant(0.0)
        thousand = g.add_constant_pos_real(1000.0)
        # overall_mean ~ Normal(0, 1000)
        overall_mean_dist = g.add_distribution(
            graph.DistributionType.NORMAL, graph.AtomicType.REAL, [zero, thousand]
        )
        overall_mean = g.add_operator(graph.OperatorType.SAMPLE, [overall_mean_dist])
        # overall_std ~ HalfCauchy(1000)
        # [note: the original paper had overall_std ~ Uniform(0, 1000)]
        overall_std_dist = g.add_distribution(
            graph.DistributionType.HALF_CAUCHY, graph.AtomicType.POS_REAL, [thousand]
        )
        overall_std = g.add_operator(graph.OperatorType.SAMPLE, [overall_std_dist])
        # for each school we will add two random variables,
        # but first we need to define a distribution
        school_effect_dist = g.add_distribution(
            graph.DistributionType.NORMAL,
            graph.AtomicType.REAL,
            [overall_mean, overall_std],
        )
        for treatment_mean_value, treatment_std_value in DATA:
            # school_effect ~ Normal(overall_mean, overall_std)
            school_effect = g.add_operator(
                graph.OperatorType.SAMPLE, [school_effect_dist]
            )
            g.query(school_effect)
            # treatment_mean ~ Normal(school_effect, treatment_std)
            treatment_std = g.add_constant_pos_real(treatment_std_value)
            treatment_mean_dist = g.add_distribution(
                graph.DistributionType.NORMAL,
                graph.AtomicType.REAL,
                [school_effect, treatment_std],
            )
            treatment_mean = g.add_operator(
                graph.OperatorType.SAMPLE, [treatment_mean_dist]
            )
            g.observe(treatment_mean, treatment_mean_value)
        g.query(overall_mean)
        g.query(overall_std)
        means = g.infer_mean(1000, graph.InferenceType.NMC)
        for idx, (mean, std) in enumerate(EXPECTED):
            self.assertTrue(
                abs(means[idx] - mean) < std * 0.5,
                f"index {idx} expected {mean} +- {std*0.5} actual {means[idx]}",
            )

    # see https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Bivariate_case
    # we are assuming zero mean here for simplicity
    def test_bivariate_gaussian(self):
        g = graph.Graph()
        flat = g.add_distribution(
            graph.DistributionType.FLAT, graph.AtomicType.REAL, []
        )
        x = g.add_operator(graph.OperatorType.SAMPLE, [flat])
        y = g.add_operator(graph.OperatorType.SAMPLE, [flat])
        x_sq = g.add_operator(graph.OperatorType.MULTIPLY, [x, x])
        y_sq = g.add_operator(graph.OperatorType.MULTIPLY, [y, y])
        x_y = g.add_operator(graph.OperatorType.MULTIPLY, [x, y])
        SIGMA_X = 5.0
        SIGMA_Y = 2.0
        RHO = 0.7
        x_sq_term = g.add_constant(-0.5 / (1 - RHO ** 2) / SIGMA_X ** 2)
        g.add_factor(graph.FactorType.EXP_PRODUCT, [x_sq, x_sq_term])
        y_sq_term = g.add_constant(-0.5 / (1 - RHO ** 2) / SIGMA_Y ** 2)
        g.add_factor(graph.FactorType.EXP_PRODUCT, [y_sq, y_sq_term])
        x_y_term = g.add_constant(RHO / (1 - RHO ** 2) / SIGMA_X / SIGMA_Y)
        g.add_factor(graph.FactorType.EXP_PRODUCT, [x_y, x_y_term])
        g.query(x)
        g.query(x_sq)
        g.query(y)
        g.query(y_sq)
        g.query(x_y)
        means = g.infer_mean(10000, graph.InferenceType.NMC)
        print("means", means)  # only printed on error
        self.assertTrue(abs(means[0] - 0.0) < 0.2, "mean of x should be 0")
        self.assertTrue(
            abs(means[1] - SIGMA_X ** 2) < 0.2, f"mean of x^2 should be {SIGMA_X**2}"
        )
        self.assertTrue(abs(means[2] - 0.0) < 0.2, "mean of y should be 0")
        self.assertTrue(
            abs(means[3] - SIGMA_Y ** 2) < 0.2, f"mean of y^2 should be {SIGMA_Y**2}"
        )
        post_cov = means[4] / math.sqrt(means[1]) / math.sqrt(means[3])
        self.assertTrue(
            abs(post_cov - RHO) < 0.2, f"covariance should be {RHO} is {post_cov}"
        )

    def test_probit_regression(self):
        """
        x ~ Normal(0, 1)
        y ~ Bernoulli(Phi(x))
        P(Phi(x) | y = true) ~ Beta(2, 1)
        """
        g = graph.Graph()
        zero = g.add_constant(0.0)
        one = g.add_constant_pos_real(1.0)
        prior = g.add_distribution(
            graph.DistributionType.NORMAL, graph.AtomicType.REAL, [zero, one]
        )
        x = g.add_operator(graph.OperatorType.SAMPLE, [prior])
        phi_x = g.add_operator(graph.OperatorType.PHI, [x])
        likelihood = g.add_distribution(
            graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [phi_x]
        )
        y = g.add_operator(graph.OperatorType.SAMPLE, [likelihood])
        g.observe(y, True)
        phi_x_sq = g.add_operator(graph.OperatorType.MULTIPLY, [phi_x, phi_x])
        g.query(phi_x)
        g.query(phi_x_sq)
        means = g.infer_mean(10000, graph.InferenceType.NMC)
        post_var = means[1] - means[0] ** 2
        self.assertAlmostEquals(
            means[0], 2 / (2 + 1), 2, f"posterior mean {means[0]} is not accurate"
        )
        self.assertAlmostEquals(
            post_var,
            2 * 1 / (2 + 1) ** 2 / (2 + 1 + 1),
            2,
            f"posterior variance {post_var} is not accurate",
        )

    def test_clara_gp(self):
        """
        CLARA-GP model
        f() ~ GP(0, squared_exp_covar)
        for each labeler l:
            spec_l ~ Beta(SPEC_ALPHA, SPEC_BETA)
            sens_l ~ Beta(SENS_ALPHA, SENS_BETA)
        for each item i
            violating_i ~ Bernoulli(Phi(f(i)))
            for each labeler l
                if violating_i
                    prob_i_l = sens_l
                else
                    prob_i_l = 1 - spec_l
                label_i_l ~ Bernoulli(prob_i_l)
        """
        ALPHA = 1.0
        RHO = 0.1
        SENS_ALPHA = 9.0
        SENS_BETA = 1.0
        SPEC_ALPHA = 9.5
        SPEC_BETA = 0.5
        NUM_LABELERS = 2
        SCORES = np.array([0.1, 0.2, 0.3])
        ITEM_LABELS = [[False, False], [False, True], [True, True]]
        # see https://mc-stan.org/docs/2_19/functions-reference/covariance.html for
        # a reference on this covariance function
        covar = ALPHA ** 2 * np.exp(
            -((np.expand_dims(SCORES, 1) - SCORES) ** 2) / 2 / RHO ** 2
        )
        tau = np.linalg.inv(covar)  # the precision matrix
        g = graph.Graph()
        # first we will create f ~ GP
        flat = g.add_distribution(
            graph.DistributionType.FLAT, graph.AtomicType.REAL, []
        )
        f = [g.add_operator(graph.OperatorType.SAMPLE, [flat]) for _ in SCORES]
        for i in range(len(SCORES)):
            tau_i_i = g.add_constant(-0.5 * tau[i, i])
            g.add_factor(graph.FactorType.EXP_PRODUCT, [tau_i_i, f[i], f[i]])
            for j in range(i + 1, len(SCORES)):
                tau_i_j = g.add_constant(-1.0 * tau[i, j])
                g.add_factor(graph.FactorType.EXP_PRODUCT, [tau_i_j, f[i], f[j]])
        # for each labeler l:
        #     spec_l ~ Beta(SPEC_ALPHA, SPEC_BETA)
        #     sens_l ~ Beta(SENS_ALPHA, SENS_BETA)
        spec_alpha = g.add_constant_pos_real(SPEC_ALPHA)
        spec_beta = g.add_constant_pos_real(SPEC_BETA)
        spec_prior = g.add_distribution(
            graph.DistributionType.BETA,
            graph.AtomicType.PROBABILITY,
            [spec_alpha, spec_beta],
        )
        sens_alpha = g.add_constant_pos_real(SENS_ALPHA)
        sens_beta = g.add_constant_pos_real(SENS_BETA)
        sens_prior = g.add_distribution(
            graph.DistributionType.BETA,
            graph.AtomicType.PROBABILITY,
            [sens_alpha, sens_beta],
        )
        spec, comp_spec, sens = [], [], []
        for labeler in range(NUM_LABELERS):
            spec.append(g.add_operator(graph.OperatorType.SAMPLE, [spec_prior]))
            comp_spec.append(
                g.add_operator(graph.OperatorType.COMPLEMENT, [spec[labeler]])
            )
            sens.append(g.add_operator(graph.OperatorType.SAMPLE, [sens_prior]))
        # for each item i
        for i, labels in enumerate(ITEM_LABELS):
            # violating_i ~ Bernoulli(Phi(f(i)))
            dist_i = g.add_distribution(
                graph.DistributionType.BERNOULLI,
                graph.AtomicType.BOOLEAN,
                [g.add_operator(graph.OperatorType.PHI, [f[i]])],
            )
            violating_i = g.add_operator(graph.OperatorType.SAMPLE, [dist_i])
            # for each labeler l
            for lidx, label_val in enumerate(labels):
                # if violating_i
                #     prob_i_l = sens_l
                # else
                #     prob_i_l = 1 - spec_l
                prob_i_l = g.add_operator(
                    graph.OperatorType.IF_THEN_ELSE,
                    [violating_i, sens[lidx], comp_spec[lidx]],
                )
                # label_i_l ~ Bernoulli(prob_i_l)
                dist_i_l = g.add_distribution(
                    graph.DistributionType.BERNOULLI,
                    graph.AtomicType.BOOLEAN,
                    [prob_i_l],
                )
                label_i_l = g.add_operator(graph.OperatorType.SAMPLE, [dist_i_l])
                g.observe(label_i_l, label_val)
            g.query(violating_i)
        means = g.infer_mean(1000, graph.InferenceType.NMC)
        self.assertLess(means[0], means[1])
        self.assertLess(means[1], means[2])

    def test_uncoupled_bools(self):
        """
        X_1 ~ Bernoulli(0.5)
        X_2 ~ Bernoulli(0.5)
        P(X_1 == X_2) = 0.5
        """
        g = graph.Graph()
        half = g.add_constant_probability(0.5)
        bernoulli = g.add_distribution(
            graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [half]
        )
        X_1 = g.add_operator(graph.OperatorType.SAMPLE, [bernoulli])
        X_2 = g.add_operator(graph.OperatorType.SAMPLE, [bernoulli])
        g.query(X_1)
        g.query(X_2)
        prob_equal = (
            sum(x == y for (x, y) in g.infer(100000, graph.InferenceType.NMC)) / 100000
        )
        self.assertAlmostEqual(prob_equal, 0.5, delta=0.01)

    def test_coupled_bools(self):
        """
        X_1 ~ Bernoulli(0.5)
        X_2 ~ Bernoulli(0.5)
        sigma_1 = 1 if X_1 else -1
        sigma_2 = 1 if X_2 else -1
        target += sigma_1 * sigma_2
        P(X_1 == X_2) = e / (e + e^-1)
        """
        g = graph.Graph()
        half = g.add_constant_probability(0.5)
        bernoulli = g.add_distribution(
            graph.DistributionType.BERNOULLI, graph.AtomicType.BOOLEAN, [half]
        )
        X_1 = g.add_operator(graph.OperatorType.SAMPLE, [bernoulli])
        X_2 = g.add_operator(graph.OperatorType.SAMPLE, [bernoulli])
        plus_one = g.add_constant(1.0)
        minus_one = g.add_constant(-1.0)
        sigma_1 = g.add_operator(
            graph.OperatorType.IF_THEN_ELSE, [X_1, plus_one, minus_one]
        )
        sigma_2 = g.add_operator(
            graph.OperatorType.IF_THEN_ELSE, [X_2, plus_one, minus_one]
        )
        g.add_factor(graph.FactorType.EXP_PRODUCT, [sigma_1, sigma_2])
        g.query(X_1)
        g.query(X_2)
        prob_equal = (
            sum(x == y for (x, y) in g.infer(100000, graph.InferenceType.NMC)) / 100000
        )
        self.assertAlmostEqual(prob_equal, 0.88, delta=0.01)

    @classmethod
    def create_GPfactor(cls, bmg, alpha, rho, scores, mu=0.0):
        # see https://mc-stan.org/docs/2_19/functions-reference/covariance.html for
        # a reference on this covariance function
        covar = alpha ** 2 * np.exp(
            -((np.expand_dims(scores, 1) - scores) ** 2) / 2 / rho ** 2
        )
        tau = np.linalg.inv(covar)  # the precision matrix
        neg_mu = bmg.add_constant(-mu)
        # f ~ GP
        flat = bmg.add_distribution(
            graph.DistributionType.FLAT, graph.AtomicType.REAL, []
        )
        f = [bmg.add_operator(graph.OperatorType.SAMPLE, [flat]) for _ in scores]
        if mu == 0.0:
            f_centered = f
        else:
            f_centered = [
                bmg.add_operator(graph.OperatorType.ADD, [fi, neg_mu]) for fi in f
            ]
        for i in range(len(scores)):
            tau_i_i = bmg.add_constant(-0.5 * tau[i, i])
            bmg.add_factor(
                graph.FactorType.EXP_PRODUCT, [tau_i_i, f_centered[i], f_centered[i]]
            )
            for j in range(i + 1, len(scores)):
                tau_i_j = bmg.add_constant(-1.0 * tau[i, j])
                bmg.add_factor(
                    graph.FactorType.EXP_PRODUCT,
                    [tau_i_j, f_centered[i], f_centered[j]],
                )
        return f

    @classmethod
    def sum_negate_nodes(cls, bmg, in_nodes):
        result = bmg.add_operator(
            graph.OperatorType.NEGATE,
            [
                bmg.add_operator(
                    graph.OperatorType.TO_REAL,
                    [bmg.add_operator(graph.OperatorType.ADD, in_nodes)],
                )
            ],
        )
        return result

    def test_clara_gp_logit(self):
        """
        CLARA-GP model with prev, sens, spec in logit space
        f_prev() ~ GP(0, squared_exp_covar)
        f_sens() ~ GP(logit(0.9), squared_exp_covar)
        f_spec() ~ GP(logit(0.95), squared_exp_covar)
        for each item i
            log_prev_i = -log1pexp(-f_prev(i)) # log(prev_i)
            log_comp_prev_i = -log1pexp(f_prev(i)) # log(1 - prev_i)
            # assume all labeller share the same sens and spec
            # so sens and spec only depends on score, indexed by i
            log_spec_i = -log1pexp(-f_spec(i))
            log_com_spec_i = -log1pexp(f_spec(i))
            log_sens_i = -log1pexp(-f_sens(i))
            log_comp_sens_i = -log1pexp(f_sens(i))
            loglik1, loglik2 = log_prev_i, log_comp_prev_i
            for each label
                loglik1 += label_i_l ? log_sens_i : log_comp_sens_i
                loglik2 += label_i_l ? log_comp_spec_i : log_spec_i
            add factor:
                logsumexp(loglik1, loglk2)
        """
        ALPHA = 1.0
        RHO = 0.1
        SPEC_MU = 2.9  # logit(0.95)
        SENS_MU = 2.2  # logit(0.9)
        # NUM_LABELERS = 2
        SCORES = np.array([0.1, 0.2, 0.3])
        ITEM_LABELS = [[False, False], [False, True], [True, True]]
        # create f ~ GP
        g = graph.Graph()
        f_prev = self.create_GPfactor(g, ALPHA, RHO, SCORES)
        f_spec = self.create_GPfactor(g, ALPHA, RHO, SCORES, SPEC_MU)
        f_sens = self.create_GPfactor(g, ALPHA, RHO, SCORES, SENS_MU)
        # for each factor:
        #   -log(p) = lop1pexp(-f)
        #   -log(1-p) = log1pexp(f)
        # note: the followings log_* are negative log probabilities,
        #   negate right before LOGSUMEXP

        # for each item i
        for i, labels in enumerate(ITEM_LABELS):
            # in this test case, we assume labelers share the same spec and sens
            log_spec = g.add_operator(
                graph.OperatorType.LOG1PEXP,
                [g.add_operator(graph.OperatorType.NEGATE, [f_spec[i]])],
            )
            log_comp_spec = g.add_operator(graph.OperatorType.LOG1PEXP, [f_spec[i]])
            log_sens = g.add_operator(
                graph.OperatorType.LOG1PEXP,
                [g.add_operator(graph.OperatorType.NEGATE, [f_sens[i]])],
            )
            log_comp_sens = g.add_operator(graph.OperatorType.LOG1PEXP, [f_sens[i]])
            log_prev = g.add_operator(
                graph.OperatorType.LOG1PEXP,
                [g.add_operator(graph.OperatorType.NEGATE, [f_prev[i]])],
            )
            log_comp_prev = g.add_operator(graph.OperatorType.LOG1PEXP, [f_prev[i]])
            loglik1, loglik2 = [log_prev], [log_comp_prev]
            # for each labeler l
            for label_val in labels:
                if label_val:
                    loglik1.append(log_sens)
                    loglik2.append(log_comp_spec)
                else:
                    loglik1.append(log_comp_sens)
                    loglik2.append(log_spec)
            loglik1 = self.sum_negate_nodes(g, loglik1)
            loglik2 = self.sum_negate_nodes(g, loglik2)
            g.add_factor(
                graph.FactorType.EXP_PRODUCT,
                [g.add_operator(graph.OperatorType.LOGSUMEXP, [loglik1, loglik2])],
            )
            g.query(f_prev[i])
        means = g.infer_mean(1000, graph.InferenceType.NMC)
        self.assertLess(means[0], means[1])
        self.assertLess(means[1], means[2])
